docs/guides/add-onboarding-flow.mdx

```diff
@@ -0,0 +1,232 @@
+
+## Configure your Next.js middleware to read session data
+
+Clerk's authMiddleware() allows you ton configure access to your routes with fine grained control. You can also retrieve claims directly from the session and redirect your user accordingly. [Learn More](https://clerk.com/docs/references/nextjs/auth-middleware#usage)
+
+Add the code sample below to your `src/middleware.ts` file:
+
+```tsx
+import { authMiddleware } from "@clerk/nextjs";
+import { redirectToSignIn } from "@clerk/nextjs/server";
+import { NextRequest, NextResponse } from "next/server";
+
+// This example protects all routes including api/trpc routes
+// Please edit this to allow other routes to be public as needed.
+// See https://clerk.com/docs/references/nextjs/auth-middleware for more information about configuring your Middleware
+export default authMiddleware({
+  publicRoutes: ["/"],
+  afterAuth: async (auth, req: NextRequest) => {
+    const { userId, sessionClaims } = auth;
+
+    // For user visiting /onboarding, don't try and redirect
+    if (userId && req.nextUrl.pathname === "/onboarding") {
+      return NextResponse.next();
+    }
+
+    // User isn't signed in and the route is private -- redirect to sign-in
+    if (!userId && !auth.isPublicRoute)
+      return redirectToSignIn({ returnBackUrl: req.url });
+
+    // Catch users who doesn't have `onboardingComplete: true` in PublicMetata
+    // Redirect them to the /onboading out to complete onboarding
+    if (userId && !sessionClaims?.metadata?.onboardingComplete) {
+      const onboardingUrl = new URL("/onboarding", req.url);
+      return NextResponse.redirect(onboardingUrl);
+    }
+
+    // User is logged in and the route is protected - let them view.
+    if (userId && !auth.isPublicRoute) return NextResponse.next();
+
+    // If the route is public, anyone can view it.
+    if (auth.isPublicRoute) return NextResponse.next();
+  },
+});
+
+export const config = {
+  matcher: ["/((?!.+\\.[\\w]+$|_next).*)", "/", "/(api|trpc)(.*)"],
+};
+```
+
+Next, create a `layout.tsx` file in `src/app/onboarding` and add the following code to the file. This logic could go in the Middleware, but by adding to the `layout.tsx` to the route the logic remains in one place. This file can also be expanded to handle multiple steps, if multiple steps are required for an onboarding flow.
+
+```tsx
+import { auth } from "@clerk/nextjs/server";
+import { redirect } from "next/navigation";
+
+export default function RootLayout({
+  children,
+}: {
+  children: React.ReactNode;
+}) {
+  // Check if a user has completed onboarding
+  // If yes, redirect them to /dashboard
+  if (auth().sessionClaims?.metadata.onboardingComplete === true) {
+    redirect("/dashboard");
+  }
+
+  return <>{children}</>;
+}
+```
+
+Now that we have the logic for where to direct the user, we’ll need a way to track their onboarding status and note it on their session, let’s dig into that now!
```

Comment:
Is this removed in favor to the changes made in "Update publicMetadata based on onboarding state"

Reply:
I remove this text that's at the end of every step, because it's redundant. The user can simply move on to the next step to see what they need to do next. Maybe in blogs this is okay, because blogs are more informal and the writer speaks directly to the user more, but in technical writing, this is not proper prose.